### ISSUE TITLE
Movement refactorization

### ISSUE THREAD

**Issue by raphlinus** [google/xi-editor#159]
Right now, the movement commands are zero-parameter notifications, with names derived from Cocoa selectors (the best place to see the list is probably the [rpc dispatching](https://github.com/google/xi-editor/blob/master/rust/core-lib/src/rpc.rs#L161-L189). I'd like to refactor it, for a number of reasons, not least of which is the number of requests.

Most pure movement requests come in two forms: move the cursor, and extend selection. At the very least, I'd like to reduce the duplication. There are other reasons to do this. I'm planning to implement multiple cursors soon, and I'd like to factor out the logic to compute the movement for a single cursor from the loop that iterates over all the cursors. Also on the horizon is vi binding support, which also separates movement from other editing logic.

Here are some choices, in no particular order:

1. Add "extend selection" boolean parameter to existing movement commands. This is closest to the existing implementation.

2. Have "move" and "extend_selection" commands, which take a "movement" string-enum parameter, with values like "up", "beginning_of_document", etc. I'm thinking there can also be a "delete" command. However, in many cases the delete logic has special differences from pure movement. Two examples: delete_to_end_of_paragraph sets the kill ring, and delete_backward has finer granularity than left/right arrow keys (it can remove the accent from a composite character).

3. Same as proposal 2, but further group pairs such as left/right by having an additional parameter indicating the sign, maybe an int which is +1 or -1. It wouldn't have to be constrained to that either, for example "9j" in vi would move 9 lines down. One tricky part is naming; "-1 line" and "+1 line" obviously mean "up" and "down", but how to distinguish "beginning of line" and "left end of line" in a way that takes a sign parameter?

4. Similar to proposal 2, but have a further breakdown of unit and extent. The unit could be grapheme cluster, visible line (ie taking breaks into account), source line, buffer. The extent could be "previous" ,"next", "left", "right", "beginning", "end". There could also be a repeat count. Not all combinations would make sense; for example, beginning and end of grapheme cluster would be meaningless because the cursor is always on a grapheme cluster boundary. The concept of "beginning of line" also has some subtlety, because it could be before or after leading whitespace, depending on the context (in Mac Sublime Text, ⌘← moves to the beginning after leading whitespace, unless it is already there, in which case it moves to the absolute beginning).

I'm leaning toward 2 because I think it's the simplest, but am open to discussion. What do people think?

**Comment by cmyr** [comment]
Thinking about future vi(m) support seems useful here. I'm thinking particularly about visual mode. In this case, a movement command becomes "extend selection" based on the editor's mode. Otherwise it's based on some key modifier, e.g  ← is a movement command and ⇪← is a selection command. 

In the second case, could we also not think of this as being two actions, a modal "begin selection"/"visual mode" action and then a movement action?  This probably gets hairy quickly, because all of a sudden we're tracking state in the front-end, which we probably want to avoid as much as possible...

The granularity of deletion is an interesting observation, and also a tricky implementation detail. Should deleting backwards only remove a diacritical from a grapheme cluster that was inserted as a composition of two graphemes, e.g. on macOS? If I'm typing on a french keyboard and I type an é, even if it is inserted decomposed I wouldn't want deleting backwards to only remove the accent. There are lots of other weird edge cases. I don't fully understand the delete_to_end_of_paragraph example; I would have thought that either all delete commands yank or else none do. Do you have a clipboard model in mind, generally?

Another consideration on thinking forward to vi support: is there the possibility in the future for describing (in a plugin, maybe) a new text object, and having that be part of a movement command? I was just playing around in xi and I see that the `moveWordLeft:` selector (⌥←) is current unimplemented (`Error: Unknown edit method`). Can I specify movement by word? By method body? 

(I like modal editors, and my thinking here is definitely biased from that perspective).

I think it is reasonable that a movement or a selection (and possibly a deletion) command should be able to encapsulate the range of such commands in vi. To that end, I would want `8j`, `viw`, `vb5`, `G`, etcetera to be expressible in the command system. 

So I guess my bias is towards 4? It seems like representing these sorts of commands with option 2 will be cumbersome, but that might just be a failure of my understanding. I prefer 2 otherwise.

I also think that the concern about the beginning/end of grapheme clusters is a bit of a red herring: if we were only supporting ASCII we could make the same complaint about characters; at any rate it doesn't seem like a big issue.

Forgive me if I'm missing anything obvious; I appreciate your doing your thinking out loud, and I look forward to seeing how this comes  together. 😄

**Comment by raphlinus** [comment]
A few very brief comments:

I haven't implemented word selection yet, mostly because it wasn't clear to me exactly where word boundaries should be. Implementing the Unicode algorithm is quite nontrivial. But I think I'll do it soon, based on a simple model with spaces and ASCII punctuation.

Method bodies (and other things, like matching brackets), require the cooperation of a plugin, as these are semantic concepts specific to a particular language. I do want to support them, but a bit further down the road - a much higher priority is just getting some kind of auto-indentation.

To be very clear, what I'm talking about here is editing commands sent by the front-end to the core. It's likely that vi commands would go through a separate entry point, something like "execute_vi_command". A major reason for that is that handling of vi commands should be (more or less) consistent between all front ends, while mapping of arrow keys to commands is front end specific. That said, an important goal is to make the implementation clean and simple, so I'd want to have GUI and vi commands converge to the same implementation after a relatively simple decoding of the input.

Regarding the kill behavior, I'm just copying the existing standard platform behavior on macOS. On Mac, there's a separate kill ring (scoped to the app, shared between windows within the app) and pasteboard (system-global). Other systems might be slightly different.

Thanks for the comments. I appreciate being able to talk these things out, and I think it'll lead to a better final result.

**Comment by rkusa** [comment]
Is there a scenario where the cursor is moved without changing the selection? Maybe I am missing something, but I cannot come up with such an scenario. That is, I would also suggest to not duplicate "move" and "extend_selection" movements, and have something like a "start_selection" and "end_selection" command instead, which in the end puts the backend into a visual mode where the normal movement commands still apply.

I'd also like to note that something like "next/previous line" does also have subtlety too, because it could either apply to a physical line or a display line (e.g. j/k vs gj/gk in vim).

**Comment by cmyr** [comment]
@raphlinus Interesting stuff on backspace, I hadn't thought much about non-latin scripts. I can think of a number of situations where you wouldn't want to be deleting single codepoints, though, such as removing variation selectors from HKJ characters, or when dealing with the various combining marks, or with zero-width-joiners... obviously there's a lot of room for bike-shedding on this one, so I'm happy to let you figure out something that makes sense. I'm sure the android implementation is well thought out.

On segmentation: there's a native Rust unicode-segmentation implementation at https://github.com/unicode-rs/unicode-segmentation, which might be a reasonable place to start for word segmentation. Last I checked there were a few issues with some of the emoji + ZWJ changes in Unicode 9, but they aren't egregious.

**Comment by raphlinus** [comment]
Yes, unicode-segmentation is a good place to start, but the current interface has two major problems. First, it only iterates forward, and one of the main things we need is to determine the _previous_ boundary from a cursor position. Second, it's only implemented for `&str`, and can't deal with non-contiguous string representations. Both can be fixed, but I'll want to talk with the maintainers of that crate to see how much they want done there, and what should be in a separate crate (like xi-unicode).

That said, the unicode word boundaries, while fine for prose, are probably not what you want for code. I haven't dug into it deeply, but I do see that sublime has a "word_separators" property. Among things not too like about the pure unicode solution is that it puts word boundaries in the middle of multi-symbol operators like "::" and "->".

**Comment by cmyr** [comment]
Good points. Are you still pursuing https://github.com/google/fancy-regex / do you think that might ultimately be the basis of some of this? (I assume that project comes froma desire to support more nuanced syntax highlighting? I read this in some issue somewhere..)

I guess another option is computing and caching word boundaries (maybe in the line cache?), which would let you avoid the look-back problem at the expense of some overhead. Not sure if this is a sane idea or not.

**Comment by st-pasha** [comment]
There is one use case which appears to be in the middle ground between single and multiple cursors: the bidi cursor. "Bidi" means bidirectional, and it is used for dealing with right-to-left languages (such as arabic, hebrew, urdu, etc). With bidi cursor, there are 2 visible cursors on the screen, but when you enter a character, it will appear only at one of those cursors, depending on whether the character entered was RTL or LTR.

A description of a possible implementation (from an end-user perspective) can be found here: https://ux.stackexchange.com/questions/36774/designing-bidi-text-editing
